File: java21/src/main/java/io/github/sps23/designpatterns/observer/MarketTicker.java
package io.github.sps23.designpatterns.observer;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public final class MarketTicker {
    public record PriceUpdate(String symbol, BigDecimal price, Instant timestamp) {
    }

    @FunctionalInterface
    public interface PriceObserver {
        void onPriceUpdate(PriceUpdate update);
    }

    private final Set<PriceObserver> observers = ConcurrentHashMap.newKeySet();
    private final Map<String, BigDecimal> latestPrices = new ConcurrentHashMap<>();

    public void subscribe(PriceObserver observer) {
        observers.add(observer);
    }

    public void unsubscribe(PriceObserver observer) {
        observers.remove(observer);
    }

    public PriceUpdate publishPrice(String symbol, BigDecimal price) {
        var update = new PriceUpdate(symbol, price, Instant.now());
        latestPrices.put(symbol, price);
        observers.forEach(observer -> observer.onPriceUpdate(update));
        return update;
    }

    public Optional<BigDecimal> latestPrice(String symbol) {
        return Optional.ofNullable(latestPrices.get(symbol));
    }
}

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/observer/MarketTicker.kt
package io.github.sps23.designpatterns.observer

import java.math.BigDecimal
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

class MarketTicker {
    data class PriceUpdate(
        val symbol: String,
        val price: BigDecimal,
        val timestamp: Instant,
    )

    fun interface PriceObserver {
        fun onPriceUpdate(update: PriceUpdate)
    }

    private val observers = mutableSetOf<PriceObserver>()
    private val latestPrices = ConcurrentHashMap<String, BigDecimal>()

    fun subscribe(observer: PriceObserver) {
        observers += observer
    }

    fun unsubscribe(observer: PriceObserver) {
        observers -= observer
    }

    fun publishPrice(
        symbol: String,
        price: BigDecimal,
    ): PriceUpdate {
        val update = PriceUpdate(symbol, price, Instant.now())
        latestPrices[symbol] = price
        observers.toList().forEach { it.onPriceUpdate(update) }
        return update
    }

    fun latestPrice(symbol: String): BigDecimal? = latestPrices[symbol]
}
